{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Muta3DMaps(v 1.0) Usage\n",
    "\n",
    "## Import Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"/data/zzf/Work/SIFTS_Plus_Muta_Maps/src/py_src/\")\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from RetrievePDB import *\n",
    "from UniProt_unit import UniProt_unit\n",
    "from MMCIFplus import *\n",
    "from SIFTS_unit import SIFTS_unit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.\t建立数据集\n",
    "\n",
    "### 1.1 UniProt ID Mapping数据集的创建\n",
    "\n",
    "在本软件的适用任务中，常有需要将序列上的突变位点信息与三维结构数据里的位点对应上。当序列的ID标识符不是UniProt数据库的UniProt ID时，例如RefSeq的转录本ID标识符，就需要将这类ID标识符先进行转换。\n",
    "\n",
    "#### 1.1.1 输入数据样例\n",
    "\n",
    "下面使用范例数据进行演示："
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_unp</th>\n",
       "      <th>GENE</th>\n",
       "      <th>RefSeq_protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K45E</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NP_689699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P293A</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NP_689699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G76S</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>NP_940978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>N105I</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>NP_940978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A375S</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>NP_940978</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  mutation_unp    GENE RefSeq_protein\n",
       "0         K45E  SAMD11      NP_689699\n",
       "1        P293A  SAMD11      NP_689699\n",
       "2         G76S    AGRN      NP_940978\n",
       "3        N105I    AGRN      NP_940978\n",
       "4        A375S    AGRN      NP_940978"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 所有的文件IO都在该文件夹下进行\n",
    "folder = \"/home/zzf/Work/demo_files/Muta3dMaps/\"\n",
    "demo_file = folder + \"demo_muta.tsv\"\n",
    "demo_df = pd.read_csv(demo_file, sep='\\t')\n",
    "demo_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> 在此范例数据中，位点信息依附于特定RefSeq ID标识符上，且有对应基因\n",
    "\n",
    "#### 1.1.2\t目标列及对应信息\n",
    "\n",
    "依据范例数据的情况来指定下列参数:\n",
    "* ```id_col```: RefSeq_protein\n",
    "* ```id_type```: Refseq Protein -> P_REFSEQ_AC\n",
    "* ```muta_col```: mutation_unp\n",
    "* ```gene_col```: GENE\n",
    "* ```muta_type```: mutation in UniProt Site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# specify column names\n",
    "id_col = 'RefSeq_protein'\n",
    "id_type = 'P_REFSEQ_AC'\n",
    "muta_col = 'mutation_unp'\n",
    "gene_col = 'GENE'\n",
    "\n",
    "# Necessary Columns for ID Mapping\n",
    "usecols = ['id', 'genes', 'reviewed', 'comment(ALTERNATIVE%20PRODUCTS)', 'organism', 'protein%20names']  \n",
    "\n",
    "# Report of Data Processing\n",
    "reportPath = folder + 'Report.txt' \n",
    "# OutPut File of ID Mapping (RAW)\n",
    "rawOutputPath = folder + 'rawMapping.tsv' \n",
    "# OutPut File of ID Mapping (Final Result)\n",
    "handledOutputPath = folder + 'modifiedMapping.tsv'  \n",
    "\n",
    "# Filtering data\n",
    "constraint_dict = {\n",
    "    \"GENE_status\": (False, \"ne\"),  # 'ne' for !=\n",
    "    \"Status\": (\"reviewed\", \"eq\"),  # 'eq' for ==\n",
    "    \"unp_map_tage\": (\"Untrusted & No Isoform\", \"ne\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.1.3 获取数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 初始化\n",
    "unp_demo = UniProt_unit(demo_df, id_col, id_type, usecols, reportPath, muta_col=muta_col, gene_col=gene_col)\n",
    "# Return True if get RAW Result Successfully \n",
    "unp_demo.get_raw_ID_Mapping(rawOutputPath)\n",
    "# Deal with different situations \n",
    "handled_ID_Mapping = unp_demo.handle_ID_Mapping() \n",
    "# Add Gene Status \n",
    "unp_demo.getGeneStatus(handled_ID_Mapping)\n",
    "# Label Mapping Status \n",
    "unp_demo.label_mapping_status(handled_ID_Mapping, constraint_dict) \n",
    "# close the file-handle of report \n",
    "unp_demo.report.close() \n",
    "# Output the final result \n",
    "handled_ID_Mapping.to_csv(handledOutputPath, sep='\\t', index=False) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.1.4 样例输出"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Status</th>\n",
       "      <th>Alternative products (isoforms)</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>canonical_isoform</th>\n",
       "      <th>unp_map_tage</th>\n",
       "      <th>yourlist</th>\n",
       "      <th>UniProt</th>\n",
       "      <th>GENE</th>\n",
       "      <th>GENE_status</th>\n",
       "      <th>Mapping_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q96NU1</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>ALTERNATIVE PRODUCTS:  Event=Alternative promo...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Sterile alpha motif domain-containing protein ...</td>\n",
       "      <td>Q96NU1-3</td>\n",
       "      <td>Untrusted &amp; No Isoform</td>\n",
       "      <td>NP_689699</td>\n",
       "      <td>Q96NU1</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>True</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P43489</td>\n",
       "      <td>TNFRSF4 TXGP1L</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Tumor necrosis factor receptor superfamily mem...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Trusted &amp; No Isoform</td>\n",
       "      <td>NP_003318</td>\n",
       "      <td>P43489</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>True</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024R084</td>\n",
       "      <td>SDF4 hCG_19193</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Stromal cell derived factor 4, isoform CRA_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Trusted &amp; No Isoform</td>\n",
       "      <td>NP_057260</td>\n",
       "      <td>A0A024R084</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>True</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q96L58</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Beta-1,3-galactosyltransferase 6 (Beta-1,3-Gal...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Trusted &amp; No Isoform</td>\n",
       "      <td>NP_542172</td>\n",
       "      <td>Q96L58</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>True</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A7WPU8</td>\n",
       "      <td>SRY hCG_1733150</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Sex-determining region Y protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Trusted &amp; No Isoform</td>\n",
       "      <td>NP_003131</td>\n",
       "      <td>A7WPU8</td>\n",
       "      <td>SRY</td>\n",
       "      <td>True</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Q05066</td>\n",
       "      <td>SRY TDF</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Sex-determining region Y protein (Testis-deter...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Trusted &amp; No Isoform</td>\n",
       "      <td>NP_003131</td>\n",
       "      <td>Q05066</td>\n",
       "      <td>SRY</td>\n",
       "      <td>True</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A024R189</td>\n",
       "      <td>TBL1Y hCG_1997773</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Transducin (Beta)-like 1Y-linked, isoform CRA_a</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Trusted &amp; No Isoform</td>\n",
       "      <td>NP_150600</td>\n",
       "      <td>A0A024R189</td>\n",
       "      <td>TBL1Y</td>\n",
       "      <td>True</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q9BQ87</td>\n",
       "      <td>TBL1Y TBL1</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>F-box-like/WD repeat-containing protein TBL1Y ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Trusted &amp; No Isoform</td>\n",
       "      <td>NP_150600</td>\n",
       "      <td>Q9BQ87</td>\n",
       "      <td>TBL1Y</td>\n",
       "      <td>True</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A0A024R9E7</td>\n",
       "      <td>NLGN4Y hCG_1988457</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Neuroligin 4, Y-linked, isoform CRA_d</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Trusted &amp; No Isoform</td>\n",
       "      <td>NP_055708</td>\n",
       "      <td>A0A024R9E7</td>\n",
       "      <td>NLGN4Y</td>\n",
       "      <td>True</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>O00468</td>\n",
       "      <td>AGRN AGRIN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>ALTERNATIVE PRODUCTS:  Event=Alternative splic...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Agrin [Cleaved into: Agrin N-terminal 110 kDa ...</td>\n",
       "      <td>O00468-1</td>\n",
       "      <td>Trusted &amp; Isoform</td>\n",
       "      <td>NP_940978</td>\n",
       "      <td>O00468-6</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>True</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Q9BRK5</td>\n",
       "      <td>SDF4 CAB45 PSEC0034</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>ALTERNATIVE PRODUCTS:  Event=Alternative splic...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>45 kDa calcium-binding protein (Cab45) (Stroma...</td>\n",
       "      <td>Q9BRK5-1</td>\n",
       "      <td>Trusted &amp; Isoform</td>\n",
       "      <td>NP_057260</td>\n",
       "      <td>Q9BRK5</td>\n",
       "      <td>SDF4</td>\n",
       "      <td>True</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Q8NFZ3</td>\n",
       "      <td>NLGN4Y KIAA0951</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>ALTERNATIVE PRODUCTS:  Event=Alternative splic...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Neuroligin-4, Y-linked (Neuroligin Y)</td>\n",
       "      <td>Q8NFZ3-1</td>\n",
       "      <td>Trusted &amp; Isoform</td>\n",
       "      <td>NP_055708</td>\n",
       "      <td>Q8NFZ3</td>\n",
       "      <td>NLGN4Y</td>\n",
       "      <td>True</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Entry           Gene names      Status  \\\n",
       "0       Q96NU1               SAMD11    reviewed   \n",
       "1       P43489       TNFRSF4 TXGP1L    reviewed   \n",
       "2   A0A024R084       SDF4 hCG_19193  unreviewed   \n",
       "3       Q96L58              B3GALT6    reviewed   \n",
       "4       A7WPU8      SRY hCG_1733150  unreviewed   \n",
       "5       Q05066              SRY TDF    reviewed   \n",
       "6   A0A024R189    TBL1Y hCG_1997773  unreviewed   \n",
       "7       Q9BQ87           TBL1Y TBL1    reviewed   \n",
       "8   A0A024R9E7   NLGN4Y hCG_1988457  unreviewed   \n",
       "9       O00468           AGRN AGRIN    reviewed   \n",
       "10      Q9BRK5  SDF4 CAB45 PSEC0034    reviewed   \n",
       "11      Q8NFZ3      NLGN4Y KIAA0951    reviewed   \n",
       "\n",
       "                      Alternative products (isoforms)              Organism  \\\n",
       "0   ALTERNATIVE PRODUCTS:  Event=Alternative promo...  Homo sapiens (Human)   \n",
       "1                                                 NaN  Homo sapiens (Human)   \n",
       "2                                                 NaN  Homo sapiens (Human)   \n",
       "3                                                 NaN  Homo sapiens (Human)   \n",
       "4                                                 NaN  Homo sapiens (Human)   \n",
       "5                                                 NaN  Homo sapiens (Human)   \n",
       "6                                                 NaN  Homo sapiens (Human)   \n",
       "7                                                 NaN  Homo sapiens (Human)   \n",
       "8                                                 NaN  Homo sapiens (Human)   \n",
       "9   ALTERNATIVE PRODUCTS:  Event=Alternative splic...  Homo sapiens (Human)   \n",
       "10  ALTERNATIVE PRODUCTS:  Event=Alternative splic...  Homo sapiens (Human)   \n",
       "11  ALTERNATIVE PRODUCTS:  Event=Alternative splic...  Homo sapiens (Human)   \n",
       "\n",
       "                                        Protein names canonical_isoform  \\\n",
       "0   Sterile alpha motif domain-containing protein ...          Q96NU1-3   \n",
       "1   Tumor necrosis factor receptor superfamily mem...               NaN   \n",
       "2        Stromal cell derived factor 4, isoform CRA_c               NaN   \n",
       "3   Beta-1,3-galactosyltransferase 6 (Beta-1,3-Gal...               NaN   \n",
       "4                    Sex-determining region Y protein               NaN   \n",
       "5   Sex-determining region Y protein (Testis-deter...               NaN   \n",
       "6     Transducin (Beta)-like 1Y-linked, isoform CRA_a               NaN   \n",
       "7   F-box-like/WD repeat-containing protein TBL1Y ...               NaN   \n",
       "8               Neuroligin 4, Y-linked, isoform CRA_d               NaN   \n",
       "9   Agrin [Cleaved into: Agrin N-terminal 110 kDa ...          O00468-1   \n",
       "10  45 kDa calcium-binding protein (Cab45) (Stroma...          Q9BRK5-1   \n",
       "11              Neuroligin-4, Y-linked (Neuroligin Y)          Q8NFZ3-1   \n",
       "\n",
       "              unp_map_tage   yourlist     UniProt     GENE  GENE_status  \\\n",
       "0   Untrusted & No Isoform  NP_689699      Q96NU1   SAMD11         True   \n",
       "1     Trusted & No Isoform  NP_003318      P43489  TNFRSF4         True   \n",
       "2     Trusted & No Isoform  NP_057260  A0A024R084     SDF4         True   \n",
       "3     Trusted & No Isoform  NP_542172      Q96L58  B3GALT6         True   \n",
       "4     Trusted & No Isoform  NP_003131      A7WPU8      SRY         True   \n",
       "5     Trusted & No Isoform  NP_003131      Q05066      SRY         True   \n",
       "6     Trusted & No Isoform  NP_150600  A0A024R189    TBL1Y         True   \n",
       "7     Trusted & No Isoform  NP_150600      Q9BQ87    TBL1Y         True   \n",
       "8     Trusted & No Isoform  NP_055708  A0A024R9E7   NLGN4Y         True   \n",
       "9        Trusted & Isoform  NP_940978    O00468-6     AGRN         True   \n",
       "10       Trusted & Isoform  NP_057260      Q9BRK5     SDF4         True   \n",
       "11       Trusted & Isoform  NP_055708      Q8NFZ3   NLGN4Y         True   \n",
       "\n",
       "   Mapping_status  \n",
       "0              No  \n",
       "1             Yes  \n",
       "2              No  \n",
       "3             Yes  \n",
       "4              No  \n",
       "5             Yes  \n",
       "6              No  \n",
       "7             Yes  \n",
       "8              No  \n",
       "9             Yes  \n",
       "10            Yes  \n",
       "11            Yes  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "handled_ID_Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.1.5 Report Statistic\n",
    "\n",
    "可以看到各列数据的缺失值情况、未匹配到UniProt ID的RefSeq ID以及匹配错误数据等"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# RAW ID MAPPING FILE RESULT\n",
      "# /home/zzf/Work/demo_files/Muta3dMaps/rawMapping.tsv\n",
      "Entry                              0\n",
      "Gene names                         0\n",
      "Status                             0\n",
      "Alternative products (isoforms)    8\n",
      "Organism                           0\n",
      "Protein names                      0\n",
      "yourlist                           0\n",
      "isomap                             9\n",
      "dtype: int64\n",
      "# All id: 8\n",
      "# Unmapped id: 0\n",
      "# Untrusted id: 1\n",
      "NP_689699\n",
      "# Error id: 0\n",
      "Empty DataFrame\n",
      "Columns: [Entry, Gene names, Status, Alternative products (isoforms), Organism, Protein names, canonical_isoform, unp_map_tage, yourlist, UniProt, GENE, GENE_status, Mapping_status]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "with open(reportPath, \"rt\") as reportFile:\n",
    "    print(reportFile.read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 获取整合好的突变信息"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RefSeq_protein\n",
      "NP_003131    [I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...\n",
      "NP_003318                                               [R65C]\n",
      "NP_055708                                              [I679V]\n",
      "NP_057260                                              [D295N]\n",
      "NP_150600                                        [D69H, R176W]\n",
      "NP_542172    [R6W, L26P, V61L, S65G, P67L, T79A, D144N, D15...\n",
      "NP_689699                                        [K45E, P293A]\n",
      "NP_940978    [G76S, N105I, A375S, T435M, G510S, L1176P, R15...\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(unp_demo.muta_li)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 SIFTS数据集的创建\n",
    "\n",
    "将Uniprot Sequence(包括isoform)上的突变位点信息映射到PDB结构上，就必须进行序列比对以获得残基水平的一一对应关系。而这个工作已由SIFTS完成。\n",
    "\n",
    "SIFTS已将PDB的SEQRES序列与各UniProt Isoform进行序列比对，并且经常更新。通过调用SIFTS的[API接口](http://www.ebi.ac.uk/pdbe/api/mappings/all_isoforms/:accession)，解析JSON格式的返回数据，即可获取到PDB相关链与Uniprot Isoform的残基位置对应关系与范围以及序列相似性等信息。残基对应关系中的PDB残基号码是从1开始计数\n",
    "的SEQRES残基索引。\n",
    "\n",
    "#### 1.3.1 获取原始数据集"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "getSiftsInfo(): Start to get the pdb info from SIFTS. 1d0a\n",
      "getSiftsInfo(): End a circle.[ 1d0a ] current: 1 ALL: 12\n",
      "getSiftsInfo(): Start to get the pdb info from SIFTS. 1hry\n",
      "getSiftsInfo(): End a circle.[ 1hry ] current: 2 ALL: 12\n",
      "getSiftsInfo(): Start to get the pdb info from SIFTS. 1hrz\n",
      "getSiftsInfo(): End a circle.[ 1hrz ] current: 3 ALL: 12\n",
      "getSiftsInfo(): Start to get the pdb info from SIFTS. 1j46\n",
      "getSiftsInfo(): End a circle.[ 1j46 ] current: 4 ALL: 12\n",
      "getSiftsInfo(): Start to get the pdb info from SIFTS. 1j47\n",
      "getSiftsInfo(): End a circle.[ 1j47 ] current: 5 ALL: 12\n",
      "getSiftsInfo(): Start to get the pdb info from SIFTS. 2gzk\n",
      "getSiftsInfo(): End a circle.[ 2gzk ] current: 6 ALL: 12\n",
      "getSiftsInfo(): Start to get the pdb info from SIFTS. 2hev\n",
      "getSiftsInfo(): End a circle.[ 2hev ] current: 7 ALL: 12\n",
      "getSiftsInfo(): Start to get the pdb info from SIFTS. 2hey\n",
      "getSiftsInfo(): End a circle.[ 2hey ] current: 8 ALL: 12\n",
      "getSiftsInfo(): Start to get the pdb info from SIFTS. 6edb\n",
      "getSiftsInfo(): End a circle.[ 6edb ] current: 9 ALL: 12\n",
      "getSiftsInfo(): Start to get the pdb info from SIFTS. 6ogx\n",
      "getSiftsInfo(): End a circle.[ 6ogx ] current: 10 ALL: 12\n",
      "getSiftsInfo(): Start to get the pdb info from SIFTS. 6okm\n",
      "getSiftsInfo(): End a circle.[ 6okm ] current: 11 ALL: 12\n",
      "getSiftsInfo(): Start to get the pdb info from SIFTS. 6okn\n",
      "getSiftsInfo(): End a circle.[ 6okn ] current: 12 ALL: 12\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(0, [])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SIFTS 原始数据文件路径\n",
    "raw_sifts_file_path = 'pdb_uniprot_SIFTS_raw_demo.tsv'\n",
    "# SIFTS 加上经过整合的覆盖范围区间信息后的文件路径\n",
    "add_rangeInfo_sifts_file_path = 'pdb_uniprot_SIFTS_addRangeInfo_demo.tsv'\n",
    "# SIFTS 判断PDB链相对UniProt Isoform的序列是否有Intertion,Deletion,并加上标签\n",
    "add_InDe_sifts_file_path = 'pdb_uniprot_SIFTS_delwithInDe_demo.tsv'\n",
    "# 中间文件保存文件夹\n",
    "SIFTS_unit.CONFIG['DOWNLOAD_FOLDER'] = folder\n",
    "\n",
    "# 初始化\n",
    "sifts_demo = SIFTS_unit()\n",
    "# 获取SIFTS当前可提供UniProt-PDB映射信息的UniProt与PDB ID\n",
    "info_dict = sifts_demo.get_info_from_uniprot_pdb_file(\n",
    "    related_unp=handled_ID_Mapping[handled_ID_Mapping[\"Mapping_status\"]==\"Yes\"][\"Entry\"].to_list())\n",
    "# 设置预备获取信息的PDB ID集\n",
    "sifts_demo.pdb_list = sorted(info_dict['pdb_set'])\n",
    "# 获取SIFTS 原始数据\n",
    "sifts_demo.get_raw_SIFTS(outputPath='%s%s' % (SIFTS_unit.CONFIG['DOWNLOAD_FOLDER'], raw_sifts_file_path))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.3.2 处理原始数据集\n",
    "\n",
    "解析SIFTS提供的残基对应关系，可以间接判断出SEQRES序列相对于完整蛋白序列(Uniprot Sequence)的差异，包括Insertion，Deletion以及SEQRES序列头尾的差异部分 (SEQRES序列在头或尾具有而Uniprot序列不具有的部分)。\n",
    "\n",
    "我们将原本SIFTS提供的pdb_start, pdb_end, unp_start, unp_end 残基水平对应关系转换为区间格式，如下表sifts_pdb_range, sifts_unp_range所示。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>chain_id</th>\n",
       "      <th>UniProt</th>\n",
       "      <th>identity</th>\n",
       "      <th>identifier</th>\n",
       "      <th>is_canonical</th>\n",
       "      <th>sifts_pdb_range</th>\n",
       "      <th>sifts_unp_range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>A</td>\n",
       "      <td>Q12933</td>\n",
       "      <td>0.994</td>\n",
       "      <td>TRAF2_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>[[1, 168]]</td>\n",
       "      <td>[[334, 501]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>B</td>\n",
       "      <td>Q12933</td>\n",
       "      <td>0.994</td>\n",
       "      <td>TRAF2_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>[[1, 168]]</td>\n",
       "      <td>[[334, 501]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>C</td>\n",
       "      <td>Q12933</td>\n",
       "      <td>0.994</td>\n",
       "      <td>TRAF2_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>[[1, 168]]</td>\n",
       "      <td>[[334, 501]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>D</td>\n",
       "      <td>Q12933</td>\n",
       "      <td>0.994</td>\n",
       "      <td>TRAF2_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>[[1, 168]]</td>\n",
       "      <td>[[334, 501]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>E</td>\n",
       "      <td>Q12933</td>\n",
       "      <td>0.994</td>\n",
       "      <td>TRAF2_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>[[1, 168]]</td>\n",
       "      <td>[[334, 501]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  pdb_id chain_id UniProt  identity   identifier  is_canonical  \\\n",
       "0   1D0A        A  Q12933     0.994  TRAF2_HUMAN          True   \n",
       "1   1D0A        B  Q12933     0.994  TRAF2_HUMAN          True   \n",
       "2   1D0A        C  Q12933     0.994  TRAF2_HUMAN          True   \n",
       "3   1D0A        D  Q12933     0.994  TRAF2_HUMAN          True   \n",
       "4   1D0A        E  Q12933     0.994  TRAF2_HUMAN          True   \n",
       "\n",
       "  sifts_pdb_range sifts_unp_range  \n",
       "0      [[1, 168]]    [[334, 501]]  \n",
       "1      [[1, 168]]    [[334, 501]]  \n",
       "2      [[1, 168]]    [[334, 501]]  \n",
       "3      [[1, 168]]    [[334, 501]]  \n",
       "4      [[1, 168]]    [[334, 501]]  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 处理SIFTS原始数据: 加上经过整合的覆盖范围区间等信息\n",
    "handle_sifts_df = sifts_demo.handle_SIFTS(outputPath='%s%s' % (SIFTS_unit.CONFIG['DOWNLOAD_FOLDER'], add_rangeInfo_sifts_file_path))\n",
    "handle_sifts_df[[\"pdb_id\", \"chain_id\",\"UniProt\", \"identity\",\"identifier\", \"is_canonical\", \"sifts_pdb_range\", \"sifts_unp_range\"]].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "大多数情况下，SIFTS提供的残基水平对应关系转换为区间后只会有一个子区间，pdb与unp子区间对应的残基序列长度是相等的，我们记为Safe。但是，存在特殊情况： \n",
    "* 对应子区间的长度并不总是相等   pdb链相对于uniprot Isoform Sequence发生了Deletion \n",
    "* 有时pdb与unp的对应关系会有多个子区间   pdb链相对于uniprot Isoform Sequence发生了Insertion \n",
    "* unp子区间之间并不连续   pdb链相对于uniprot Isoform Sequence发生了Insertion 以及 Deletion，存在较长差异序列\n",
    "发现上述特殊情况中的Insertion以及Insertion & Deletion时后续可以调用以及封装好的函数进行pair-wise alignment(采用Biopyton模块的pairwise2)以确定具体的差异内容, 更正sifts_pdb_range, sifts_unp_range使得残基对应关系更为精确。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>chain_id</th>\n",
       "      <th>UniProt</th>\n",
       "      <th>yourlist</th>\n",
       "      <th>identity</th>\n",
       "      <th>sifts_pdb_range</th>\n",
       "      <th>sifts_unp_range</th>\n",
       "      <th>delete</th>\n",
       "      <th>sifts_range_tage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>G</td>\n",
       "      <td>P43489</td>\n",
       "      <td>NP_003318</td>\n",
       "      <td>0.833</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>False</td>\n",
       "      <td>Safe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>H</td>\n",
       "      <td>P43489</td>\n",
       "      <td>NP_003318</td>\n",
       "      <td>0.833</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>False</td>\n",
       "      <td>Safe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>I</td>\n",
       "      <td>P43489</td>\n",
       "      <td>NP_003318</td>\n",
       "      <td>0.833</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>False</td>\n",
       "      <td>Safe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>J</td>\n",
       "      <td>P43489</td>\n",
       "      <td>NP_003318</td>\n",
       "      <td>0.833</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>False</td>\n",
       "      <td>Safe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>K</td>\n",
       "      <td>P43489</td>\n",
       "      <td>NP_003318</td>\n",
       "      <td>0.833</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>False</td>\n",
       "      <td>Safe</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  pdb_id chain_id UniProt   yourlist  identity sifts_pdb_range  \\\n",
       "0   1D0A        G  P43489  NP_003318     0.833        [[1, 6]]   \n",
       "1   1D0A        H  P43489  NP_003318     0.833        [[1, 6]]   \n",
       "2   1D0A        I  P43489  NP_003318     0.833        [[1, 6]]   \n",
       "3   1D0A        J  P43489  NP_003318     0.833        [[1, 6]]   \n",
       "4   1D0A        K  P43489  NP_003318     0.833        [[1, 6]]   \n",
       "\n",
       "  sifts_unp_range  delete sifts_range_tage  \n",
       "0    [[261, 266]]   False             Safe  \n",
       "1    [[261, 266]]   False             Safe  \n",
       "2    [[261, 266]]   False             Safe  \n",
       "3    [[261, 266]]   False             Safe  \n",
       "4    [[261, 266]]   False             Safe  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 处理SIFTS原始数据: 判断PDB链相对UniProt Isoform的序列是否有Intertion,Deletion,并加上标签\n",
    "sifts_demo.deal_with_insertionDeletion_SIFTS(sifts_df=handle_sifts_df, outputPath='%s%s' % (SIFTS_unit.CONFIG['DOWNLOAD_FOLDER'], add_InDe_sifts_file_path))\n",
    "# 将RefSeq ID 信息进行整合\n",
    "handle_sifts_df = pd.merge(handle_sifts_df, handled_ID_Mapping[['UniProt','yourlist']].drop_duplicates())\n",
    "handle_sifts_df[[\"pdb_id\", \"chain_id\", \"UniProt\", \"yourlist\", \"identity\",\"sifts_pdb_range\", \"sifts_unp_range\", \"delete\",\"sifts_range_tage\"]].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.4 PDB文件的下载\n",
    "\n",
    "#### 1.4.1 样例脚本"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RetrievePDB: {downloadPath: /home/zzf/Work/demo_files/Muta3dMaps/MMCIF/, format: mmCIF, tail: .cif, raw_tail: .cif, prefix: , ftpSite: RCSB, host: ftp.rcsb.org, dividedPath: pub/pdb/data/structures/divided, len(pdbs): 0, len(fail): 0, }\n",
      "220---------- Welcome to Pure-FTPd [privsep] [TLS] ----------\n",
      "220-You are user number 2 of 500 allowed.\n",
      "220-Local time is now 08:20. Server port: 2421.\n",
      "220-Only anonymous FTP is allowed here\n",
      "220 You will be disconnected after 15 minutes of inactivity.\n",
      "Downloading File: /home/zzf/Work/demo_files/Muta3dMaps/MMCIF/1D0A.cif.gz\n",
      "Downloading File: /home/zzf/Work/demo_files/Muta3dMaps/MMCIF/6EDB.cif.gz\n",
      "Downloading File: /home/zzf/Work/demo_files/Muta3dMaps/MMCIF/2GZK.cif.gz\n",
      "Downloading File: /home/zzf/Work/demo_files/Muta3dMaps/MMCIF/2HEV.cif.gz\n",
      "Downloading File: /home/zzf/Work/demo_files/Muta3dMaps/MMCIF/2HEY.cif.gz\n",
      "Downloading File: /home/zzf/Work/demo_files/Muta3dMaps/MMCIF/1HRY.cif.gz\n",
      "Downloading File: /home/zzf/Work/demo_files/Muta3dMaps/MMCIF/1HRZ.cif.gz\n",
      "Downloading File: /home/zzf/Work/demo_files/Muta3dMaps/MMCIF/1J46.cif.gz\n",
      "Downloading File: /home/zzf/Work/demo_files/Muta3dMaps/MMCIF/1J47.cif.gz\n",
      "Downloading File: /home/zzf/Work/demo_files/Muta3dMaps/MMCIF/6OGX.cif.gz\n",
      "Downloading File: /home/zzf/Work/demo_files/Muta3dMaps/MMCIF/6OKM.cif.gz\n",
      "Downloading File: /home/zzf/Work/demo_files/Muta3dMaps/MMCIF/6OKN.cif.gz\n",
      "FTP Closed\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "# 预下载的PDB\n",
    "pdbs = handle_sifts_df['pdb_id'].drop_duplicates()\n",
    "# 下载路径\n",
    "path = folder + \"MMCIF/\"\n",
    "# 默认下载MMCIF格式\n",
    "mpw = MPWrapper(path)\n",
    "fail = mpw.ftp_retrieve_batch(pdbs)\n",
    "# 查看下载失败的PDB ID集\n",
    "print(fail)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['.ipynb_checkpoints', '1D0A.cif', '6EDB.cif', '2GZK.cif', '2HEV.cif', '2HEY.cif', '1HRY.cif', '1HRZ.cif', '1J46.cif', '1J47.cif', '6OGX.cif', '6OKM.cif', '6OKN.cif']\n"
     ]
    }
   ],
   "source": [
    "print(os.listdir(path))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.5 MMCIF文件信息的获取与整合\n",
    "\n",
    "#### 1.5.1 建立MMCIF信息数据集"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RetrievePDB: {downloadPath: /home/zzf/Work/demo_files/Muta3dMaps/MMCIF/, format: mmCIF, tail: .cif, raw_tail: .cif, prefix: , ftpSite: RCSB, host: ftp.rcsb.org, dividedPath: pub/pdb/data/structures/divided, len(pdbs): 0, len(fail): 0, }\n",
      "data_1D0A\n",
      "data_2HEV\n",
      "data_2HEY\n",
      "data_6OGX\n",
      "data_6OKM\n",
      "data_6OKN\n",
      "data_1HRY\n",
      "data_1HRZ\n",
      "data_1J46\n",
      "data_1J47\n",
      "data_2GZK\n",
      "data_6EDB\n",
      "handle_mmcif_data(): Modified Dict\n",
      "handle_mmcif_data(): Modified Dfrm\n"
     ]
    }
   ],
   "source": [
    "# 设置文件路径\n",
    "raw_mmcif_path = folder + 'rawMMCIF2Dfrm.tsv'\n",
    "handled_mmcif_path = folder + 'handledMMCIF2Dfrm.tsv'\n",
    "MMCIF_FILE_FOLDER['MMCIF_NEW_FOLDER'] = path\n",
    "\n",
    "# 初始化\n",
    "mmcif_demo = MMCIF2Dfrm()\n",
    "# 检测pdbs文件路径,若未下载相关MMCIF文件则会自动下载\n",
    "mmcif_demo.check_mmcif_file(pdbs)\n",
    "# 接续\n",
    "if os.path.exists(handled_mmcif_path):\n",
    "    finished_li = set(pd.read_csv(handled_mmcif_path, sep='\\t', usecols=['pdb_id'])['pdb_id'])\n",
    "else:\n",
    "    finished_li = []\n",
    "\n",
    "mmcif_demo.update_mmcif_result(raw_mmcif_path, handled_mmcif_path, finished=finished_li)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.5.2 范例输出"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entity_id</th>\n",
       "      <th>protein_type</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>chain_id</th>\n",
       "      <th>_pdbx_poly_seq_scheme.mon_id</th>\n",
       "      <th>_pdbx_poly_seq_scheme.pdb_mon_id</th>\n",
       "      <th>_pdbx_poly_seq_scheme.auth_mon_id</th>\n",
       "      <th>_pdbx_poly_seq_scheme.ndb_seq_num</th>\n",
       "      <th>_pdbx_poly_seq_scheme.pdb_seq_num</th>\n",
       "      <th>_pdbx_poly_seq_scheme.auth_seq_num</th>\n",
       "      <th>...</th>\n",
       "      <th>mutation_num</th>\n",
       "      <th>metal_ligand_num</th>\n",
       "      <th>Modification_num</th>\n",
       "      <th>seqres_len</th>\n",
       "      <th>coordinates_len</th>\n",
       "      <th>Modification_index</th>\n",
       "      <th>mis_index</th>\n",
       "      <th>mis_range</th>\n",
       "      <th>resolution_score</th>\n",
       "      <th>protein_chain_and_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>1D0A</td>\n",
       "      <td>A</td>\n",
       "      <td>AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;1...</td>\n",
       "      <td>334;335;336;337;338;339;340;341;342;343;344;34...</td>\n",
       "      <td>334;335;336;337;338;339;340;341;342;343;344;34...</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>168</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>1D0A</td>\n",
       "      <td>B</td>\n",
       "      <td>AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;1...</td>\n",
       "      <td>334;335;336;337;338;339;340;341;342;343;344;34...</td>\n",
       "      <td>334;335;336;337;338;339;340;341;342;343;344;34...</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>168</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>1D0A</td>\n",
       "      <td>C</td>\n",
       "      <td>AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>AAAALAAAVLAMEASTYDGVFIWKISDFARKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;1...</td>\n",
       "      <td>334;335;336;337;338;339;340;341;342;343;344;34...</td>\n",
       "      <td>334;335;336;337;338;339;340;341;342;343;344;34...</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>168</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>1D0A</td>\n",
       "      <td>D</td>\n",
       "      <td>AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;1...</td>\n",
       "      <td>334;335;336;337;338;339;340;341;342;343;344;34...</td>\n",
       "      <td>334;335;336;337;338;339;340;341;342;343;344;34...</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>168</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>polypeptide(L)</td>\n",
       "      <td>1D0A</td>\n",
       "      <td>E</td>\n",
       "      <td>AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...</td>\n",
       "      <td>1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;1...</td>\n",
       "      <td>334;335;336;337;338;339;340;341;342;343;344;34...</td>\n",
       "      <td>334;335;336;337;338;339;340;341;342;343;344;34...</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>168</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   entity_id    protein_type pdb_id chain_id  \\\n",
       "0          1  polypeptide(L)   1D0A        A   \n",
       "1          1  polypeptide(L)   1D0A        B   \n",
       "2          1  polypeptide(L)   1D0A        C   \n",
       "3          1  polypeptide(L)   1D0A        D   \n",
       "4          1  polypeptide(L)   1D0A        E   \n",
       "\n",
       "                        _pdbx_poly_seq_scheme.mon_id  \\\n",
       "0  AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...   \n",
       "1  AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...   \n",
       "2  AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...   \n",
       "3  AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...   \n",
       "4  AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...   \n",
       "\n",
       "                    _pdbx_poly_seq_scheme.pdb_mon_id  \\\n",
       "0  AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...   \n",
       "1  AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...   \n",
       "2  AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...   \n",
       "3  AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...   \n",
       "4  AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...   \n",
       "\n",
       "                   _pdbx_poly_seq_scheme.auth_mon_id  \\\n",
       "0  AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...   \n",
       "1  AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...   \n",
       "2  AAAALAAAVLAMEASTYDGVFIWKISDFARKRQEAVAGRIPAIFSP...   \n",
       "3  AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...   \n",
       "4  AMADLEQKVLEMEASTYDGVFIWKISDFPRKRQEAVAGRIPAIFSP...   \n",
       "\n",
       "                   _pdbx_poly_seq_scheme.ndb_seq_num  \\\n",
       "0  1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;1...   \n",
       "1  1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;1...   \n",
       "2  1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;1...   \n",
       "3  1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;1...   \n",
       "4  1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16;17;18;1...   \n",
       "\n",
       "                   _pdbx_poly_seq_scheme.pdb_seq_num  \\\n",
       "0  334;335;336;337;338;339;340;341;342;343;344;34...   \n",
       "1  334;335;336;337;338;339;340;341;342;343;344;34...   \n",
       "2  334;335;336;337;338;339;340;341;342;343;344;34...   \n",
       "3  334;335;336;337;338;339;340;341;342;343;344;34...   \n",
       "4  334;335;336;337;338;339;340;341;342;343;344;34...   \n",
       "\n",
       "                  _pdbx_poly_seq_scheme.auth_seq_num  ... mutation_num  \\\n",
       "0  334;335;336;337;338;339;340;341;342;343;344;34...  ...            0   \n",
       "1  334;335;336;337;338;339;340;341;342;343;344;34...  ...            0   \n",
       "2  334;335;336;337;338;339;340;341;342;343;344;34...  ...            0   \n",
       "3  334;335;336;337;338;339;340;341;342;343;344;34...  ...            0   \n",
       "4  334;335;336;337;338;339;340;341;342;343;344;34...  ...            0   \n",
       "\n",
       "  metal_ligand_num  Modification_num seqres_len coordinates_len  \\\n",
       "0                0                 0        168             168   \n",
       "1                0                 0        168             168   \n",
       "2                0                 0        168             168   \n",
       "3                0                 0        168             168   \n",
       "4                0                 0        168             168   \n",
       "\n",
       "  Modification_index  mis_index mis_range  resolution_score  \\\n",
       "0                 []         []       NaN               2.0   \n",
       "1                 []         []       NaN               2.0   \n",
       "2                 []         []       NaN               2.0   \n",
       "3                 []         []       NaN               2.0   \n",
       "4                 []         []       NaN               2.0   \n",
       "\n",
       "                            protein_chain_and_length  \n",
       "0  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...  \n",
       "1  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...  \n",
       "2  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...  \n",
       "3  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...  \n",
       "4  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...  \n",
       "\n",
       "[5 rows x 43 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "handled_mmcif_df = pd.read_csv(handled_mmcif_path, sep=\"\\t\", converters={'chain_id':str, 'asym_id':str, 'entity_id':int})\n",
    "handled_mmcif_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2 整合SIFTS数据与PDB(MMCIF格式)数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>chain_id</th>\n",
       "      <th>UniProt</th>\n",
       "      <th>identity</th>\n",
       "      <th>identifier</th>\n",
       "      <th>is_canonical</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>entity_id</th>\n",
       "      <th>struct_asym_id</th>\n",
       "      <th>...</th>\n",
       "      <th>mutation_num</th>\n",
       "      <th>metal_ligand_num</th>\n",
       "      <th>Modification_num</th>\n",
       "      <th>seqres_len</th>\n",
       "      <th>coordinates_len</th>\n",
       "      <th>Modification_index</th>\n",
       "      <th>mis_index</th>\n",
       "      <th>mis_range</th>\n",
       "      <th>resolution_score</th>\n",
       "      <th>protein_chain_and_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>G</td>\n",
       "      <td>P43489</td>\n",
       "      <td>0.833</td>\n",
       "      <td>TNR4_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>{\"author_residue_number\": 261, \"author_inserti...</td>\n",
       "      <td>{\"author_residue_number\": 266, \"author_inserti...</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>H</td>\n",
       "      <td>P43489</td>\n",
       "      <td>0.833</td>\n",
       "      <td>TNR4_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>{\"author_residue_number\": 261, \"author_inserti...</td>\n",
       "      <td>{\"author_residue_number\": 266, \"author_inserti...</td>\n",
       "      <td>2</td>\n",
       "      <td>H</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>I</td>\n",
       "      <td>P43489</td>\n",
       "      <td>0.833</td>\n",
       "      <td>TNR4_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>{\"author_residue_number\": 261, \"author_inserti...</td>\n",
       "      <td>{\"author_residue_number\": 266, \"author_inserti...</td>\n",
       "      <td>2</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>J</td>\n",
       "      <td>P43489</td>\n",
       "      <td>0.833</td>\n",
       "      <td>TNR4_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>{\"author_residue_number\": 261, \"author_inserti...</td>\n",
       "      <td>{\"author_residue_number\": 266, \"author_inserti...</td>\n",
       "      <td>2</td>\n",
       "      <td>J</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>K</td>\n",
       "      <td>P43489</td>\n",
       "      <td>0.833</td>\n",
       "      <td>TNR4_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>{\"author_residue_number\": 261, \"author_inserti...</td>\n",
       "      <td>{\"author_residue_number\": 266, \"author_inserti...</td>\n",
       "      <td>2</td>\n",
       "      <td>K</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  pdb_id chain_id UniProt  identity  identifier  is_canonical  \\\n",
       "0   1D0A        G  P43489     0.833  TNR4_HUMAN          True   \n",
       "1   1D0A        H  P43489     0.833  TNR4_HUMAN          True   \n",
       "2   1D0A        I  P43489     0.833  TNR4_HUMAN          True   \n",
       "3   1D0A        J  P43489     0.833  TNR4_HUMAN          True   \n",
       "4   1D0A        K  P43489     0.833  TNR4_HUMAN          True   \n",
       "\n",
       "                                               start  \\\n",
       "0  {\"author_residue_number\": 261, \"author_inserti...   \n",
       "1  {\"author_residue_number\": 261, \"author_inserti...   \n",
       "2  {\"author_residue_number\": 261, \"author_inserti...   \n",
       "3  {\"author_residue_number\": 261, \"author_inserti...   \n",
       "4  {\"author_residue_number\": 261, \"author_inserti...   \n",
       "\n",
       "                                                 end  entity_id  \\\n",
       "0  {\"author_residue_number\": 266, \"author_inserti...          2   \n",
       "1  {\"author_residue_number\": 266, \"author_inserti...          2   \n",
       "2  {\"author_residue_number\": 266, \"author_inserti...          2   \n",
       "3  {\"author_residue_number\": 266, \"author_inserti...          2   \n",
       "4  {\"author_residue_number\": 266, \"author_inserti...          2   \n",
       "\n",
       "  struct_asym_id  ... mutation_num metal_ligand_num Modification_num  \\\n",
       "0              G  ...            0                1                1   \n",
       "1              H  ...            0                1                1   \n",
       "2              I  ...            0                1                1   \n",
       "3              J  ...            0                1                1   \n",
       "4              K  ...            0                1                1   \n",
       "\n",
       "  seqres_len coordinates_len  Modification_index  mis_index  mis_range  \\\n",
       "0          6               6                 [0]         []        NaN   \n",
       "1          6               6                 [0]         []        NaN   \n",
       "2          6               6                 [0]         []        NaN   \n",
       "3          6               6                 [0]         []        NaN   \n",
       "4          6               6                 [0]         []        NaN   \n",
       "\n",
       "   resolution_score                           protein_chain_and_length  \n",
       "0               2.0  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...  \n",
       "1               2.0  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...  \n",
       "2               2.0  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...  \n",
       "3               2.0  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...  \n",
       "4               2.0  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...  \n",
       "\n",
       "[5 rows x 62 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sifts_mmcif_df = sifts_demo.add_mmcif_info_SIFTS(sifts_df=handle_sifts_df, mmcif_df=handled_mmcif_df)\n",
    "sifts_mmcif_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3 将突变从UniProt映射至PDB链\n",
    "\n",
    "### 3.1 对Deletion相关数据进行修正\n",
    "\n",
    "发现上述特殊情况中的Insertion以及Insertion & Deletion时后续可以调用以及封装好的函数进行pair-wise alignment(采用Biopyton模块的pairwise2)以确定具体的差异内容, 更正sifts_pdb_range, sifts_unp_range使得残基对应关系更为精确"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>chain_id</th>\n",
       "      <th>UniProt</th>\n",
       "      <th>identity</th>\n",
       "      <th>identifier</th>\n",
       "      <th>is_canonical</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>entity_id</th>\n",
       "      <th>struct_asym_id</th>\n",
       "      <th>...</th>\n",
       "      <th>Modification_num</th>\n",
       "      <th>seqres_len</th>\n",
       "      <th>coordinates_len</th>\n",
       "      <th>Modification_index</th>\n",
       "      <th>mis_index</th>\n",
       "      <th>mis_range</th>\n",
       "      <th>resolution_score</th>\n",
       "      <th>protein_chain_and_length</th>\n",
       "      <th>new_sifts_unp_range</th>\n",
       "      <th>new_sifts_pdb_range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>G</td>\n",
       "      <td>P43489</td>\n",
       "      <td>0.833</td>\n",
       "      <td>TNR4_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>{\"author_residue_number\": 261, \"author_inserti...</td>\n",
       "      <td>{\"author_residue_number\": 266, \"author_inserti...</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>H</td>\n",
       "      <td>P43489</td>\n",
       "      <td>0.833</td>\n",
       "      <td>TNR4_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>{\"author_residue_number\": 261, \"author_inserti...</td>\n",
       "      <td>{\"author_residue_number\": 266, \"author_inserti...</td>\n",
       "      <td>2</td>\n",
       "      <td>H</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>I</td>\n",
       "      <td>P43489</td>\n",
       "      <td>0.833</td>\n",
       "      <td>TNR4_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>{\"author_residue_number\": 261, \"author_inserti...</td>\n",
       "      <td>{\"author_residue_number\": 266, \"author_inserti...</td>\n",
       "      <td>2</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>J</td>\n",
       "      <td>P43489</td>\n",
       "      <td>0.833</td>\n",
       "      <td>TNR4_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>{\"author_residue_number\": 261, \"author_inserti...</td>\n",
       "      <td>{\"author_residue_number\": 266, \"author_inserti...</td>\n",
       "      <td>2</td>\n",
       "      <td>J</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1D0A</td>\n",
       "      <td>K</td>\n",
       "      <td>P43489</td>\n",
       "      <td>0.833</td>\n",
       "      <td>TNR4_HUMAN</td>\n",
       "      <td>True</td>\n",
       "      <td>{\"author_residue_number\": 261, \"author_inserti...</td>\n",
       "      <td>{\"author_residue_number\": 266, \"author_inserti...</td>\n",
       "      <td>2</td>\n",
       "      <td>K</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>[0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  pdb_id chain_id UniProt  identity  identifier  is_canonical  \\\n",
       "0   1D0A        G  P43489     0.833  TNR4_HUMAN          True   \n",
       "1   1D0A        H  P43489     0.833  TNR4_HUMAN          True   \n",
       "2   1D0A        I  P43489     0.833  TNR4_HUMAN          True   \n",
       "3   1D0A        J  P43489     0.833  TNR4_HUMAN          True   \n",
       "4   1D0A        K  P43489     0.833  TNR4_HUMAN          True   \n",
       "\n",
       "                                               start  \\\n",
       "0  {\"author_residue_number\": 261, \"author_inserti...   \n",
       "1  {\"author_residue_number\": 261, \"author_inserti...   \n",
       "2  {\"author_residue_number\": 261, \"author_inserti...   \n",
       "3  {\"author_residue_number\": 261, \"author_inserti...   \n",
       "4  {\"author_residue_number\": 261, \"author_inserti...   \n",
       "\n",
       "                                                 end  entity_id  \\\n",
       "0  {\"author_residue_number\": 266, \"author_inserti...          2   \n",
       "1  {\"author_residue_number\": 266, \"author_inserti...          2   \n",
       "2  {\"author_residue_number\": 266, \"author_inserti...          2   \n",
       "3  {\"author_residue_number\": 266, \"author_inserti...          2   \n",
       "4  {\"author_residue_number\": 266, \"author_inserti...          2   \n",
       "\n",
       "  struct_asym_id  ... Modification_num seqres_len coordinates_len  \\\n",
       "0              G  ...                1          6               6   \n",
       "1              H  ...                1          6               6   \n",
       "2              I  ...                1          6               6   \n",
       "3              J  ...                1          6               6   \n",
       "4              K  ...                1          6               6   \n",
       "\n",
       "  Modification_index mis_index  mis_range  resolution_score  \\\n",
       "0                [0]        []        NaN               2.0   \n",
       "1                [0]        []        NaN               2.0   \n",
       "2                [0]        []        NaN               2.0   \n",
       "3                [0]        []        NaN               2.0   \n",
       "4                [0]        []        NaN               2.0   \n",
       "\n",
       "                            protein_chain_and_length  new_sifts_unp_range  \\\n",
       "0  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...         [[261, 266]]   \n",
       "1  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...         [[261, 266]]   \n",
       "2  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...         [[261, 266]]   \n",
       "3  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...         [[261, 266]]   \n",
       "4  [(168, \"A\"), (168, \"B\"), (168, \"C\"), (168, \"D\"...         [[261, 266]]   \n",
       "\n",
       "   new_sifts_pdb_range  \n",
       "0             [[1, 6]]  \n",
       "1             [[1, 6]]  \n",
       "2             [[1, 6]]  \n",
       "3             [[1, 6]]  \n",
       "4             [[1, 6]]  \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unp_fasta_files_path = folder + '/fasta_files/%s.fasta'\n",
    "update_sifts_mmcif_df = sifts_demo.update_range_info_SIFTS(unp_fasta_files_path, sifts_df=sifts_mmcif_df)\n",
    "update_sifts_mmcif_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 映射\n",
    "\n",
    "#### 3.2.1 范例脚本"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 将突变数据整合\n",
    "update_sifts_mmcif_df['mutation_unp'] = update_sifts_mmcif_df.apply(lambda x: unp_demo.muta_li[x['yourlist']], axis=1)\n",
    "# PDB突变位点信息\n",
    "muta_info_li = []\n",
    "update_sifts_mmcif_df['mutation_pdb'] = update_sifts_mmcif_df.apply(lambda x: SIFTS_unit.map_muta_from_unp_to_pdb(x, 'mutation_unp', 'new_sifts_unp_range', 'new_sifts_pdb_range', muta_info_li, unp_fasta_files_path) if not isinstance(x['new_sifts_pdb_range'], float) and not isinstance(x['mutation_unp'], float) else np.nan, axis=1)\n",
    "# 映射情况信息\n",
    "update_sifts_mmcif_df['muta_map_info'] = pd.Series(muta_info_li, index=update_sifts_mmcif_df.dropna(subset=['mutation_unp']).index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3.2.2 范例输出\n",
    "\n",
    "对于下面的输出数据，可以看到突变数据从RefSeq到PDB三维结构位点的映射情况，参考```muta_map_info```列可得知具体映射结果：\n",
    "* Safe: 突变位点成功映射到PDB三维结构上，且氨基酸残基正确对应\n",
    "* PossibleMutation: 突变位点成功映射到PDB三维结构上，但氨基酸残基错误对应\n",
    "* EntityMutation:  突变位点成功映射到PDB三维结构上，但氨基酸残基错误对应，且可知对应PDB链上该位点本身存在突变数据\n",
    "* Missing：突变位点成功映射到PDB三维结构上，但是该位点缺失空间坐标信息\n",
    "* Unmapped #：突变位点没有映射到PDB三维结构上"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>yourlist</th>\n",
       "      <th>UniProt</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>chain_id</th>\n",
       "      <th>identity</th>\n",
       "      <th>new_sifts_pdb_range</th>\n",
       "      <th>new_sifts_unp_range</th>\n",
       "      <th>sifts_range_tage</th>\n",
       "      <th>mutation_unp</th>\n",
       "      <th>mutation_pdb</th>\n",
       "      <th>muta_map_info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_003318</td>\n",
       "      <td>P43489</td>\n",
       "      <td>1D0A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.833</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[R65C]</td>\n",
       "      <td>[#]</td>\n",
       "      <td>[Unmapped #: R65C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_003318</td>\n",
       "      <td>P43489</td>\n",
       "      <td>1D0A</td>\n",
       "      <td>H</td>\n",
       "      <td>0.833</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[R65C]</td>\n",
       "      <td>[#]</td>\n",
       "      <td>[Unmapped #: R65C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_003318</td>\n",
       "      <td>P43489</td>\n",
       "      <td>1D0A</td>\n",
       "      <td>I</td>\n",
       "      <td>0.833</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[R65C]</td>\n",
       "      <td>[#]</td>\n",
       "      <td>[Unmapped #: R65C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NP_003318</td>\n",
       "      <td>P43489</td>\n",
       "      <td>1D0A</td>\n",
       "      <td>J</td>\n",
       "      <td>0.833</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[R65C]</td>\n",
       "      <td>[#]</td>\n",
       "      <td>[Unmapped #: R65C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP_003318</td>\n",
       "      <td>P43489</td>\n",
       "      <td>1D0A</td>\n",
       "      <td>K</td>\n",
       "      <td>0.833</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[R65C]</td>\n",
       "      <td>[#]</td>\n",
       "      <td>[Unmapped #: R65C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NP_003318</td>\n",
       "      <td>P43489</td>\n",
       "      <td>1D0A</td>\n",
       "      <td>L</td>\n",
       "      <td>0.833</td>\n",
       "      <td>[[1, 6]]</td>\n",
       "      <td>[[261, 266]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[R65C]</td>\n",
       "      <td>[#]</td>\n",
       "      <td>[Unmapped #: R65C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NP_003318</td>\n",
       "      <td>P43489</td>\n",
       "      <td>2HEV</td>\n",
       "      <td>R</td>\n",
       "      <td>1.000</td>\n",
       "      <td>[[5, 146]]</td>\n",
       "      <td>[[29, 170]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[R65C]</td>\n",
       "      <td>[65]</td>\n",
       "      <td>[Safe]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NP_003318</td>\n",
       "      <td>P43489</td>\n",
       "      <td>2HEY</td>\n",
       "      <td>R</td>\n",
       "      <td>1.000</td>\n",
       "      <td>[[5, 146]]</td>\n",
       "      <td>[[29, 170]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[R65C]</td>\n",
       "      <td>[65]</td>\n",
       "      <td>[Safe]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NP_003318</td>\n",
       "      <td>P43489</td>\n",
       "      <td>2HEY</td>\n",
       "      <td>T</td>\n",
       "      <td>1.000</td>\n",
       "      <td>[[5, 146]]</td>\n",
       "      <td>[[29, 170]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[R65C]</td>\n",
       "      <td>[65]</td>\n",
       "      <td>[Safe]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NP_003318</td>\n",
       "      <td>P43489</td>\n",
       "      <td>6OGX</td>\n",
       "      <td>G</td>\n",
       "      <td>0.929</td>\n",
       "      <td>[[22, 163]]</td>\n",
       "      <td>[[29, 170]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[R65C]</td>\n",
       "      <td>[65]</td>\n",
       "      <td>[Safe]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NP_003318</td>\n",
       "      <td>P43489</td>\n",
       "      <td>6OKM</td>\n",
       "      <td>R</td>\n",
       "      <td>0.929</td>\n",
       "      <td>[[22, 163]]</td>\n",
       "      <td>[[29, 170]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[R65C]</td>\n",
       "      <td>[65]</td>\n",
       "      <td>[Safe]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NP_003318</td>\n",
       "      <td>P43489</td>\n",
       "      <td>6OKN</td>\n",
       "      <td>E</td>\n",
       "      <td>0.929</td>\n",
       "      <td>[[22, 163]]</td>\n",
       "      <td>[[29, 170]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[R65C]</td>\n",
       "      <td>[65]</td>\n",
       "      <td>[Missing]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NP_003318</td>\n",
       "      <td>P43489</td>\n",
       "      <td>6OKN</td>\n",
       "      <td>R</td>\n",
       "      <td>0.929</td>\n",
       "      <td>[[22, 163]]</td>\n",
       "      <td>[[29, 170]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[R65C]</td>\n",
       "      <td>[65]</td>\n",
       "      <td>[Missing]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NP_003131</td>\n",
       "      <td>Q05066</td>\n",
       "      <td>1HRY</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000</td>\n",
       "      <td>[[1, 76]]</td>\n",
       "      <td>[[56, 131]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...</td>\n",
       "      <td>[13, 12, 12, 11, 10, 10, 9, 9, 7, 5, 5, 4, 3, ...</td>\n",
       "      <td>[Safe, Safe, Safe, Safe, Safe, Safe, Safe, Saf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NP_003131</td>\n",
       "      <td>Q05066</td>\n",
       "      <td>1HRZ</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000</td>\n",
       "      <td>[[1, 76]]</td>\n",
       "      <td>[[56, 131]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...</td>\n",
       "      <td>[13, 12, 12, 11, 10, 10, 9, 9, 7, 5, 5, 4, 3, ...</td>\n",
       "      <td>[Safe, Safe, Safe, Safe, Safe, Safe, Safe, Saf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NP_003131</td>\n",
       "      <td>Q05066</td>\n",
       "      <td>1J46</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000</td>\n",
       "      <td>[[2, 85]]</td>\n",
       "      <td>[[57, 140]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...</td>\n",
       "      <td>[13, 12, 12, 11, 10, 10, 9, 9, 7, 5, 5, 4, 3, ...</td>\n",
       "      <td>[Safe, Safe, Safe, Safe, Safe, Safe, Safe, Saf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NP_003131</td>\n",
       "      <td>Q05066</td>\n",
       "      <td>1J47</td>\n",
       "      <td>A</td>\n",
       "      <td>0.988</td>\n",
       "      <td>[[2, 85]]</td>\n",
       "      <td>[[57, 140]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...</td>\n",
       "      <td>[13, 12, 12, 11, 10, 10, 9, 9, 7, 5, 5, 4, 3, ...</td>\n",
       "      <td>[Safe, Safe, Safe, Safe, Safe, Safe, EntityMut...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NP_003131</td>\n",
       "      <td>Q05066</td>\n",
       "      <td>2GZK</td>\n",
       "      <td>A</td>\n",
       "      <td>0.804</td>\n",
       "      <td>[[1, 75]]</td>\n",
       "      <td>[[56, 130]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...</td>\n",
       "      <td>[13, 12, 12, 11, 10, 10, 9, 9, 7, 5, 5, 4, 3, ...</td>\n",
       "      <td>[Safe, Safe, Safe, Safe, Safe, Safe, Safe, Saf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NP_003131</td>\n",
       "      <td>Q05066</td>\n",
       "      <td>6EDB</td>\n",
       "      <td>A</td>\n",
       "      <td>0.716</td>\n",
       "      <td>[[2, 76]]</td>\n",
       "      <td>[[56, 130]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...</td>\n",
       "      <td>[81, 80, 80, 79, 78, 78, 77, 77, 75, 73, 73, 7...</td>\n",
       "      <td>[Missing, Missing, Missing, Missing, Missing, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NP_003131</td>\n",
       "      <td>Q05066</td>\n",
       "      <td>6EDB</td>\n",
       "      <td>B</td>\n",
       "      <td>0.716</td>\n",
       "      <td>[[2, 76]]</td>\n",
       "      <td>[[56, 130]]</td>\n",
       "      <td>Safe</td>\n",
       "      <td>[I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...</td>\n",
       "      <td>[13, 12, 12, 11, 10, 10, 9, 9, 7, 5, 5, 4, 3, ...</td>\n",
       "      <td>[Safe, Safe, Safe, Safe, Safe, Safe, Safe, Saf...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     yourlist UniProt pdb_id chain_id  identity new_sifts_pdb_range  \\\n",
       "0   NP_003318  P43489   1D0A        G     0.833            [[1, 6]]   \n",
       "1   NP_003318  P43489   1D0A        H     0.833            [[1, 6]]   \n",
       "2   NP_003318  P43489   1D0A        I     0.833            [[1, 6]]   \n",
       "3   NP_003318  P43489   1D0A        J     0.833            [[1, 6]]   \n",
       "4   NP_003318  P43489   1D0A        K     0.833            [[1, 6]]   \n",
       "5   NP_003318  P43489   1D0A        L     0.833            [[1, 6]]   \n",
       "6   NP_003318  P43489   2HEV        R     1.000          [[5, 146]]   \n",
       "7   NP_003318  P43489   2HEY        R     1.000          [[5, 146]]   \n",
       "8   NP_003318  P43489   2HEY        T     1.000          [[5, 146]]   \n",
       "9   NP_003318  P43489   6OGX        G     0.929         [[22, 163]]   \n",
       "10  NP_003318  P43489   6OKM        R     0.929         [[22, 163]]   \n",
       "11  NP_003318  P43489   6OKN        E     0.929         [[22, 163]]   \n",
       "12  NP_003318  P43489   6OKN        R     0.929         [[22, 163]]   \n",
       "13  NP_003131  Q05066   1HRY        A     1.000           [[1, 76]]   \n",
       "14  NP_003131  Q05066   1HRZ        A     1.000           [[1, 76]]   \n",
       "15  NP_003131  Q05066   1J46        A     1.000           [[2, 85]]   \n",
       "16  NP_003131  Q05066   1J47        A     0.988           [[2, 85]]   \n",
       "17  NP_003131  Q05066   2GZK        A     0.804           [[1, 75]]   \n",
       "18  NP_003131  Q05066   6EDB        A     0.716           [[2, 76]]   \n",
       "19  NP_003131  Q05066   6EDB        B     0.716           [[2, 76]]   \n",
       "\n",
       "   new_sifts_unp_range sifts_range_tage  \\\n",
       "0         [[261, 266]]             Safe   \n",
       "1         [[261, 266]]             Safe   \n",
       "2         [[261, 266]]             Safe   \n",
       "3         [[261, 266]]             Safe   \n",
       "4         [[261, 266]]             Safe   \n",
       "5         [[261, 266]]             Safe   \n",
       "6          [[29, 170]]             Safe   \n",
       "7          [[29, 170]]             Safe   \n",
       "8          [[29, 170]]             Safe   \n",
       "9          [[29, 170]]             Safe   \n",
       "10         [[29, 170]]             Safe   \n",
       "11         [[29, 170]]             Safe   \n",
       "12         [[29, 170]]             Safe   \n",
       "13         [[56, 131]]             Safe   \n",
       "14         [[56, 131]]             Safe   \n",
       "15         [[57, 140]]             Safe   \n",
       "16         [[57, 140]]             Safe   \n",
       "17         [[56, 130]]             Safe   \n",
       "18         [[56, 130]]             Safe   \n",
       "19         [[56, 130]]             Safe   \n",
       "\n",
       "                                         mutation_unp  \\\n",
       "0                                              [R65C]   \n",
       "1                                              [R65C]   \n",
       "2                                              [R65C]   \n",
       "3                                              [R65C]   \n",
       "4                                              [R65C]   \n",
       "5                                              [R65C]   \n",
       "6                                              [R65C]   \n",
       "7                                              [R65C]   \n",
       "8                                              [R65C]   \n",
       "9                                              [R65C]   \n",
       "10                                             [R65C]   \n",
       "11                                             [R65C]   \n",
       "12                                             [R65C]   \n",
       "13  [I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...   \n",
       "14  [I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...   \n",
       "15  [I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...   \n",
       "16  [I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...   \n",
       "17  [I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...   \n",
       "18  [I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...   \n",
       "19  [I68T, F67L, F67V, A66T, N65H, N65D, M64I, M64...   \n",
       "\n",
       "                                         mutation_pdb  \\\n",
       "0                                                 [#]   \n",
       "1                                                 [#]   \n",
       "2                                                 [#]   \n",
       "3                                                 [#]   \n",
       "4                                                 [#]   \n",
       "5                                                 [#]   \n",
       "6                                                [65]   \n",
       "7                                                [65]   \n",
       "8                                                [65]   \n",
       "9                                                [65]   \n",
       "10                                               [65]   \n",
       "11                                               [65]   \n",
       "12                                               [65]   \n",
       "13  [13, 12, 12, 11, 10, 10, 9, 9, 7, 5, 5, 4, 3, ...   \n",
       "14  [13, 12, 12, 11, 10, 10, 9, 9, 7, 5, 5, 4, 3, ...   \n",
       "15  [13, 12, 12, 11, 10, 10, 9, 9, 7, 5, 5, 4, 3, ...   \n",
       "16  [13, 12, 12, 11, 10, 10, 9, 9, 7, 5, 5, 4, 3, ...   \n",
       "17  [13, 12, 12, 11, 10, 10, 9, 9, 7, 5, 5, 4, 3, ...   \n",
       "18  [81, 80, 80, 79, 78, 78, 77, 77, 75, 73, 73, 7...   \n",
       "19  [13, 12, 12, 11, 10, 10, 9, 9, 7, 5, 5, 4, 3, ...   \n",
       "\n",
       "                                        muta_map_info  \n",
       "0                                  [Unmapped #: R65C]  \n",
       "1                                  [Unmapped #: R65C]  \n",
       "2                                  [Unmapped #: R65C]  \n",
       "3                                  [Unmapped #: R65C]  \n",
       "4                                  [Unmapped #: R65C]  \n",
       "5                                  [Unmapped #: R65C]  \n",
       "6                                              [Safe]  \n",
       "7                                              [Safe]  \n",
       "8                                              [Safe]  \n",
       "9                                              [Safe]  \n",
       "10                                             [Safe]  \n",
       "11                                          [Missing]  \n",
       "12                                          [Missing]  \n",
       "13  [Safe, Safe, Safe, Safe, Safe, Safe, Safe, Saf...  \n",
       "14  [Safe, Safe, Safe, Safe, Safe, Safe, Safe, Saf...  \n",
       "15  [Safe, Safe, Safe, Safe, Safe, Safe, Safe, Saf...  \n",
       "16  [Safe, Safe, Safe, Safe, Safe, Safe, EntityMut...  \n",
       "17  [Safe, Safe, Safe, Safe, Safe, Safe, Safe, Saf...  \n",
       "18  [Missing, Missing, Missing, Missing, Missing, ...  \n",
       "19  [Safe, Safe, Safe, Safe, Safe, Safe, Safe, Saf...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "update_sifts_mmcif_df[['yourlist', 'UniProt', 'pdb_id', 'chain_id', 'identity', 'new_sifts_pdb_range', 'new_sifts_unp_range', 'sifts_range_tage','mutation_unp', 'mutation_pdb', 'muta_map_info']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## *4 Interactome3D meta-data的获取\n",
    "\n",
    "> 该步骤与上述步骤不一定相关"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TYPE</th>\n",
       "      <th>PDB_ID</th>\n",
       "      <th>BIO_UNIT</th>\n",
       "      <th>FILENAME</th>\n",
       "      <th>group_compo</th>\n",
       "      <th>PROT</th>\n",
       "      <th>CHAIN</th>\n",
       "      <th>MODEL</th>\n",
       "      <th>SEQ_IDENT</th>\n",
       "      <th>COVERAGE</th>\n",
       "      <th>SEQ_BEGIN</th>\n",
       "      <th>SEQ_END</th>\n",
       "      <th>DOMAIN</th>\n",
       "      <th>model_len</th>\n",
       "      <th>model_range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Structure</td>\n",
       "      <td>2MSE</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A024RAV5-P02647-EXP-2mse.pdb1-B-0-A-0.pdb</td>\n",
       "      <td>A0A024RAV5_P02647</td>\n",
       "      <td>A0A024RAV5</td>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>98.4</td>\n",
       "      <td>1</td>\n",
       "      <td>185</td>\n",
       "      <td>-</td>\n",
       "      <td>185</td>\n",
       "      <td>[[1, 185]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Structure</td>\n",
       "      <td>2MSE</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A024RAV5-P10398-EXP-2mse.pdb1-B-0-D-0.pdb</td>\n",
       "      <td>A0A024RAV5_P10398</td>\n",
       "      <td>A0A024RAV5</td>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>98.4</td>\n",
       "      <td>1</td>\n",
       "      <td>185</td>\n",
       "      <td>-</td>\n",
       "      <td>185</td>\n",
       "      <td>[[1, 185]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Structure</td>\n",
       "      <td>4OV6</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A075B5G3-Q8NBP7-EXP-4ov6.pdb1-F-0-B-0.pdb</td>\n",
       "      <td>A0A075B5G3_Q8NBP7</td>\n",
       "      <td>A0A075B5G3</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>99</td>\n",
       "      <td>-</td>\n",
       "      <td>99</td>\n",
       "      <td>[[1, 99]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Structure</td>\n",
       "      <td>4OV6</td>\n",
       "      <td>2</td>\n",
       "      <td>A0A075B5G3-Q8NBP7-EXP-4ov6.pdb2-G-0-E-0.pdb</td>\n",
       "      <td>A0A075B5G3_Q8NBP7</td>\n",
       "      <td>A0A075B5G3</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>1</td>\n",
       "      <td>96</td>\n",
       "      <td>-</td>\n",
       "      <td>96</td>\n",
       "      <td>[[1, 96]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Structure</td>\n",
       "      <td>4OV6</td>\n",
       "      <td>1</td>\n",
       "      <td>A0A075B5G3-Q8NBP7-EXP-4ov6.pdb1-F-0-A-0.pdb</td>\n",
       "      <td>A0A075B5G3_Q8NBP7</td>\n",
       "      <td>A0A075B5G3</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>99</td>\n",
       "      <td>-</td>\n",
       "      <td>99</td>\n",
       "      <td>[[1, 99]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        TYPE PDB_ID BIO_UNIT                                     FILENAME  \\\n",
       "0  Structure   2MSE        1  A0A024RAV5-P02647-EXP-2mse.pdb1-B-0-A-0.pdb   \n",
       "1  Structure   2MSE        1  A0A024RAV5-P10398-EXP-2mse.pdb1-B-0-D-0.pdb   \n",
       "2  Structure   4OV6        1  A0A075B5G3-Q8NBP7-EXP-4ov6.pdb1-F-0-B-0.pdb   \n",
       "3  Structure   4OV6        2  A0A075B5G3-Q8NBP7-EXP-4ov6.pdb2-G-0-E-0.pdb   \n",
       "4  Structure   4OV6        1  A0A075B5G3-Q8NBP7-EXP-4ov6.pdb1-F-0-A-0.pdb   \n",
       "\n",
       "         group_compo        PROT CHAIN MODEL  SEQ_IDENT  COVERAGE  SEQ_BEGIN  \\\n",
       "0  A0A024RAV5_P02647  A0A024RAV5     B     0      100.0      98.4          1   \n",
       "1  A0A024RAV5_P10398  A0A024RAV5     B     0      100.0      98.4          1   \n",
       "2  A0A075B5G3_Q8NBP7  A0A075B5G3     F     0      100.0     100.0          1   \n",
       "3  A0A075B5G3_Q8NBP7  A0A075B5G3     G     0      100.0      97.0          1   \n",
       "4  A0A075B5G3_Q8NBP7  A0A075B5G3     F     0      100.0     100.0          1   \n",
       "\n",
       "   SEQ_END DOMAIN  model_len model_range  \n",
       "0      185      -        185  [[1, 185]]  \n",
       "1      185      -        185  [[1, 185]]  \n",
       "2       99      -         99   [[1, 99]]  \n",
       "3       96      -         96   [[1, 96]]  \n",
       "4       99      -         99   [[1, 99]]  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Interactome3D_unit import Interactome3D_unit\n",
    "\n",
    "interactDemo = Interactome3D_unit()\n",
    "# 原文件下载路径\n",
    "interactDemo.CONFIG['DOWNLOAD_FOLDER'] = folder\n",
    "# 获取经过处理的整合文件\n",
    "interact_df = interactDemo.get_interactions_meta(outputPath=interactDemo.CONFIG['DOWNLOAD_FOLDER']+'interactions_modified.tsv')\n",
    "interact_df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
